### ISSUE TITLE
UPNP finding invalid gateway on network with DMZ

### ISSUE THREAD

**Issue by Ploppy3** [godotengine/godot#25008]
**Godot version:**

3.1-beta.1

**OS/device including version:**

Windows 10

**Issue description:**

I would like to start by saying that I'm a beginner on computer network.

The UPNP class can't find the gateway, not sure what is going on.
Note that this [NPM package](https://www.npmjs.com/package/nat-upnp) package on a node server can successfully do port-forwarding on the Xiaomi router (more info bellow).

Here is the output produced by the bellow repro project:
```
device[0]:
- igd_our_addr:
- igd_service_type:
- igd_status: 5
- service_type: urn:schemas-upnp-org:device:InternetGatewayDevice:1
gateway:[Object:null]
no valid gateway
```

NB: IGD_STATUS_DISCONNECTED = 5

ISP Router | ➜ | Xiaomi Router | ➜ | PC
--- | --- | --- | --- | ---
DMZ set to Xiaomi Router IP |-| UPNP enabled \| static ip |-|

**Steps to reproduce:**

```godoscript
var port = 12345
var upnp = UPNP.new()
upnp.discover()
if upnp.get_gateway() and upnp.get_gateway().is_valid_gateway():
	upnp.add_port_mapping(port, port, 'Godot', 'UDP')
	upnp.add_port_mapping(port, port, 'Godot', 'TCP')
```

**Minimal reproduction project:**
[test-network.zip](https://github.com/godotengine/godot/files/2759226/test-network.zip)

**Comment by mhilbrunner** [comment]
Thanks for the report!

Can you share the router model?

Did you test the node package on the same device/OS as Godot?

**Comment by Ploppy3** [comment]
Hello, I will give all the information I omitted in the original post.

**My setup** is a standard one:

ISP Router | ➜ | PC
--- | --- | ---
UPNP enabled | - | -

- Godot UPNP: ✔
- Node UPNP: ✔

**Friend setup** is NOT a standard one:

ISP Router | ➜ | [Xiaomi WIFI Router 3G](https://en.techreviewer.de/xiaomi-router-3g-test/) | ➜ | PC
--- | --- | --- | --- | ---
DMZ set to Xiaomi Router IP |-| UPNP enabled \| static ip |-|

- Godot UPNP: ❌
- Node UPNP:  ✔

**Comment by mhilbrunner** [comment]
According to [this reddit post](https://www.reddit.com/r/godot/comments/bd1dg9/using_upnp_in_godot/) similar issues seem to exist with the [TP-Link Archer C2 AC900](https://www.tp-link.com/latam/home-networking/wifi-router/archer-c2/).

**Comment by creikey** [comment]
When I  `upnp.discover`, godot indefinitely freezes

**Comment by Razzeeyy** [comment]
Excuse me for posting here (tell me if I need to actually open a different issue for that, I'm not sure myself).

I have the problem is that `upnp.discover` doesn't detect my router as an Internet Gateway Device.
I even downloaded the upnpc executable (which also uses the miniupnpc library) to check if my router is being detected correctly, and it is.

Also weird behaviour is `upnp.discover` doesn't seem to block execution for me, I even set the wait time for 1 minute and it never blocked (even though never found a single device either).

How do I troubleshoot this further?

Information about my router:
![image](https://user-images.githubusercontent.com/912485/64476454-7255c000-d197-11e9-8973-2114cef47e40.png)

**Comment by creikey** [comment]
@Razzeeyy something like this does belong in a separate issue. If you are using Godot 3.1.1 stable, then the `upnp` library bundled with it is several versions behind the upstream master library, and there are many problems with it. Try it with `3.2` master and see if there's still an issue.

**Comment by Razzeeyy** [comment]
> @Razzeeyy something like this does belong in a separate issue. If you are using Godot 3.1.1 stable, then the `upnp` library bundled with it is several versions behind the upstream master library, and there are many problems with it. Try it with `3.2` master and see if there's still an issue.

Sorry my bad I forgot to mention, I'm actually using the 3.2 alpha nightly build.
![image](https://user-images.githubusercontent.com/912485/64485341-18e9a180-d228-11e9-971e-fe0361cf621d.png)

Also the version of executable I use to check is a upnpc build for windows from 2015 so I guess it's even more behind the upstream library but still works 😅 

P.S. I'll open a separate issue today 👌, will try to do a bit more thorough testing for a few more hours now...

**Comment by kameloov** [comment]
same issue here 
OpenGL ES 3.0 Renderer: AMD KABINI (DRM 2.50.0, 5.0.0-32-generic, LLVM 8.0.0)
device[0]:
- igd_our_addr: 
- igd_service_type: 
- igd_status: 6
- service_type: urn:schemas-upnp-org:device:InternetGatewayDevice:1
gateway:[Object:null]
no valid gateway
========================
Router model is Huawei HG531 v1

**Comment by burstina** [comment]
hello,
I got a similar issue:

   ` error = upnp.discover(2000, 2, "InternetGatewayDevice")
`
result is error 24 
**Socket error: bind, 10049**

**Comment by der-ac** [comment]
I also get error 24 when using discover with an Arris Router with the following details:

ARRIS DOCSIS 3.0 Touchstone Residential Gateway
HW_REV: 6
VENDOR: ARRIS Group, Inc.
BOOTR: 4.2.0.41
SW_REV: 9.1.103BE
MODEL: DG2470A

**Comment by Faless** [comment]
@burstina that issue was a different one. See #35953 

@der-ac do you get any other error/warning in the error/message log?

**Comment by der-ac** [comment]
I don't see any messages about it in the error log, discover() just returns a 24 error code. Is there some way I could gather more verbose output?

**Comment by Faless** [comment]
@der-ac did you enable UPNP on your router?

**Comment by der-ac** [comment]
UPNP was already enabled.

**Comment by hk-32** [comment]
I'm having the exact same issue with my Archer C20 AC750 on Windows 10. The discover command fails and returns the error 24 which apparently means a socket error. The funny thing is that I just tested the miniupnp library (that godot uses) by itself and it worked perfectly fine, detected the upnp device and successfully port forwarded. I will be testing my godot project on Android soon because I suspect that godot is having issues with connecting/sending anything to the router. Might be firewall or whatever, idk.

**Comment by Faless** [comment]
@hk-32 a generic 'socket error' (regression since `3.1.2` included) on windows was fixed in #35953 (and will be shipped with godot 3.2.1)

**Comment by hk-32** [comment]
oh right, thanks for the quick reply. Is there any way I can use it right now? like download the update somehow?

**Comment by Faless** [comment]
@hk-32 you can [build the engine from source](https://docs.godotengine.org/en/3.2/development/compiling/index.html) or wait for `3.2.1` to be released.

**Comment by Reneator** [comment]
Please redirect me to the correct router if im wrong with this. Found this ticket via reddit via google when searching for the "Godot Upnp unauthorized":

I have a fritz-box router (6590 newest OS) and it threw me a 1 (unauthorized) as result when doing the "add_port_mapping" although i had upnp active. But then i found out that you also have to enable port-unlocking for the device in general in the tab, where you can manuall add ports, you have to add the device and allow it to unlock its own ports.

After that it now works correctly!

**Comment by Faless** [comment]
Closing, as this should be solved in `3.3`
